### ISSUE TITLE
Users can specify search criteria to select particles in ParticleDiagnostics

### ISSUE THREAD

**Issue by nicolasaunai** [PHAREHUB/PHARE#89]
Writing all particles to the disk is heavy and most of the time useless because rarely the user wants to see them all. 

This issue is about adding to the pharein ParticleDiagnostics block, and to the cpp particle diagnostic, the possibility of selecting only specific particles, based on a criteria provided by the user.

the most general is the user providing a python function that gets contiguous particle vectors and returns a mask or a subset associated with the selected particles.

**Comment by PhilipDeegan** [comment]
This would be something specific during setup when you are setting the timestamps per diag too?
I guess it would need to be known before dumping, so we could filter on write

**Comment by nicolasaunai** [comment]
Yes, the way I see this is user may provide the ParticleDiagnostic block with a python function that takes a particle array and select those for dumping.

**Comment by nicolasaunai** [comment]
may be tricky if calling a python function (GIL etc.)
first pragamatic way of doing this could be to have hardcoded cpp filters that are parametrised  by python dict
